Mathlib/Computability/Language.lean

```diff
@@ -297,6 +297,65 @@ instance : KleeneAlgebra (Language α) :=
       rw [pow_succ, ←mul_assoc m l (l^n)]
       exact le_trans (le_mul_congr h le_rfl) ih }
 
+/-- Language `l.reverse` is defined as the set of words from `l` backwards. -/
+def reverse (l : Language α) : Language α := { w : List α | w.reverse ∈ l }
+
+@[simp]
+lemma reverse_mem_reverse : a.reverse ∈ l.reverse ↔ a ∈ l := by
+  show a.reverse.reverse ∈ l ↔ a ∈ l
+  rw [List.reverse_reverse]
+
+lemma reverse_zero : (0 : Language α).reverse = 0 := by
+  rfl
+
+lemma reverse_one : (1 : Language α).reverse = 1 := by
+  simp [reverse, ← one_def]
+
+lemma reverse_reverse : l.reverse.reverse = l := by
+  ext w
+  convert_to w.reverse.reverse ∈ reverse (reverse l) ↔ w ∈ l using 5
+  · rw [List.reverse_reverse]
+  rw [reverse_mem_reverse, reverse_mem_reverse]
+
+lemma reverse_union : (l + m).reverse = l.reverse + m.reverse := by
+  ext w
+  apply mem_add
+
+lemma reverse_concatenation : (l * m).reverse = m.reverse * l.reverse := by
+  ext w
+  show
+    (∃ u v, u ∈ l ∧ v ∈ m ∧ u ++ v = w.reverse) ↔
+    (∃ u v, u.reverse ∈ m ∧ v.reverse ∈ l ∧ u ++ v = w)
+  constructor <;>
+  · rintro ⟨u, v, hul, hvm, huvw⟩
+    use v.reverse, u.reverse
+    simp_all [← List.reverse_append]
+
+lemma reverse_kstar : l∗.reverse = l.reverse∗ := by
+  ext w
+  simp only [reverse, kstar_def]
+  show
+    (∃ L, w.reverse = join L ∧ ∀ y, y ∈ L → y ∈ l) ↔
+    (∃ L, w = join L ∧ ∀ y, y ∈ L → y.reverse ∈ l)
+  constructor
```

Comment:
I would guess there are some tricks to make the proof from here a lot shorter; for instance rewriting the existential through some lemma about involutive functions, and how `map f` is involutive if `f` is.

Reply:
Can you be more concrete please?

Reply:
You'll need the lemmas from #8642

Reply:
Did you just create a whole PR in order to simplify my proof?

Reply:
Yes, but also because I think those lemmas should exist.

Reply:
The outline will be to rewrite by `List.reverse_involutive.list_map.surjective.exists` or similar to reverse the `L` that is being quantified over.

Reply:
OK, I'll wait for #8642 to be merged.

Reply:
I pushed a golf